src/tabs/tokens.js

```diff
@@ -8,51 +8,89 @@ import metaversefile from '../../metaversefile-api.js';
 import {tokensHost} from '../../constants';
 import {preview} from '../../preview.js';
 
-export const Tokens = ({userOpen, hacks, address}) => {
+export const Tokens = ({userOpen, loginFrom, hacks, address}) => {
   const [nftPreviews, setNftPreviews] = useState({});
   const [nfts, setNfts] = useState(null);
   const [fetchPromises, setFetchPromises] = useState([]);
 
   useEffect(() => {
-    if (address && !nfts) {
+    if (address && !nfts && loginFrom) {
       setNfts([]);
-
       (async () => {
-        const res = await fetch(`https://api.opensea.io/api/v1/assets?owner=${address}&limit=${50}`, {
-          headers: {
-            'X-API-KEY': '6a7ceb45f3c44c84be65779ad2907046',
-          },
-        });
-        const j = await res.json();
-        const {assets} = j;
-        setNfts(assets);
-        // console.log('got assets', assets);
+        if (loginFrom === 'metamask') {
+          const res = await fetch(`https://api.opensea.io/api/v1/assets?owner=${address}&limit=${50}`, {
+            headers: {
+              'X-API-KEY': '6a7ceb45f3c44c84be65779ad2907046',
+            },
+          });
+
+          const j = await res.json();
+          const {assets} = j;
+          setNfts(assets);
+        } else if (loginFrom === 'discord') {
+          let res = await fetch(`${tokensHost}/${address}`);
+          res = await res.json();
+          res = res.map(_nft => {
+            /** Modify the response recieved from the API-Backend to match standardise format */
+            _nft.image_preview_url = _nft.hash;
+            return _nft;
+          });
+          setNfts(res);
+        }
       })();
     }
-  }, [address, nfts]);
+  }, [address, nfts, loginFrom]);
 
+  useEffect(() => {
+    if (nfts) {
+      for (const nft of nfts) {
+        if (!nftPreviews[nft.image_preview_url]) {
```

Comment:
Can you explain how this routes to totum to the local iframe or some remote server to generate the preview? What is the flow like?